sys/src/9/amd64/devvq.c

```diff
@@ -0,0 +1,595 @@
+				vqinterrupt(dev->vqs[i]);
+			}
+		}
+	}
+	return;
+}
+
+// The interrupt handler part that handles the virtqueue.
+
+static void
+vqinterrupt(Virtq *q)
+{
+	int id, free, m;
+	Rock *r;
+	Rendez *z;
+	m = q->num - 1;
+	ilock(&q->l);
+	while((q->lastused ^ q->used->idx) & m) {
+		id = q->used->ring[q->lastused++ & m].id;
+		if(r = q->rock[id]){
+			q->rock[id] = nil;
+			z = r->sleep;
+			r->done = 1;	/* hands off */
+			if(z != nil)
+				wakeup(z);
+		}
+		do {
+			free = id;
+			id = q->desc[free].next;
+			q->desc[free].next = q->free;
+			q->free = free;
+			q->nfree++;
+		} while(q->desc[free].flags & VIRTQ_DESC_F_NEXT);
+	}
+	((vqctl *)(q->pdev))->vqiomap &= ~(1 << (q->idx));	// clear bit in IO map for this queue
+	iunlock(&q->l);
+}
+
+// Generic buffer read-write operation, based on 9front sdvirtio logic. Take buffer address and length,
+// read/write direction. Get a free descriptor from the queue, fill it with the buffer information.
+// Update queue index, notify the device. The Rock structure (sleep/wakeup semaphore) is kept on the
+// process stack, and its address is stored in the queue descriptor's rock array at the index same
+// as the index of the used buffer descriptor. When the device processes the buffer, it sends an interrupt.
+// The interrupt handler scans the used descriptiors ring, and if a descriptor is found on which a process
+// is waiting, the process is awaken, and the descriptor returns to the available ring.
+// Return 0 if OK, -1 if error.
+// Q: is it secure to give interrupt handler access to the sleeping process stack? Would it be better
+// to preallocate all Rock's as part of the queue descriptor (128 bit per each vs. 64bit per a pointer).
+
+
+int
+vqrw(Virtq *q, void *a, uint64_t len, int rw)	// rw can be either 0 for read or VIRTQ_DESC_F_WRITE (2, other bits will be cleared)
+{
+	Proc *up = externup();
+	int free, head;
+	rw &= VIRTQ_DESC_F_WRITE;
+	ilock(&q->l);
+	while(q->nfree < 1) {
+		iunlock(&q->l);
+		if(!waserror())
+			tsleep(&up->sleep, return0, 0, 500);
+		poperror();
+		ilock(&q->l);
+	}
+	head = free = q->free;
+	struct virtq_desc *d = &q->desc[free];
+	free = d->next;
+	d->addr = PADDR(a);
+	d->len = len;
+	d->flags = rw;
+	Rock rock;								// the sleep-wakeup semaphore on the process stack
+	rock.done = 0;
+	rock.sleep = &up->sleep;
+	q->rock[head] = &rock;
+	q->avail->ring[q->avail->idx & (q->num - 1)] = head;
+	((vqctl *)(q->pdev))->vqiomap |= (1 << (q->idx));	// set bit in IO map for this queue
+	coherence();
+	q->avail->idx++;
+	iunlock(&q->l);
+	if((q->used->flags & VIRTQ_AVAIL_F_NO_INTERRUPT) == 0) {
+		vqctl *dev = q->pdev;
+		outs(dev->port + VIRTIO_PCI_QUEUE_NOTIFY, q->idx);
+	}
+	while(!rock.done) {
+		while(waserror())
+			;
+		tsleep(rock.sleep, viodone, &rock, 1000);
+	}
+	if(!rock.done)
+		vqinterrupt(q);
+	return 0;
+}
+
+static char *
+mapdev(vqctl *vc)
+{
+	char *dmap = nil;
+	for(int i = 0; i < nelem(dmtab) ; i++) {
+		if(vc->pci->did == dmtab[i].did) {
```

Comment:
This PR is based on another PR, or standalone, or ... what's going on here? Sorry if I am missing something.